### PR TITLE
Create simple documentation for @typeofweb/schema library

### PR DESCRIPTION
**Pull request by wisnie**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -3,3 +3,204 @@
...
+// Return 'Micheal'
+const micheal = stringValidator('Micheal');
+```
+
+### number
+
+Generates schema that will match number.
+
+```js
+const numberSchema = string();
```
Comment (mmiszy) [inline README.md:93]:
```suggestion
const numberSchema = number();
```

File: README.md [hunk 2]
```diff
@@ -3,3 +3,204 @@
...
+// Return 3.14
+const pi = numberValidator(3.14);
+```
+
+### boolean
+
+Generates schema that will match boolean.
+
+```js
+const booleanSchema = string();
```
Comment (mmiszy) [inline README.md:105]:
```suggestion
const booleanSchema = boolean();
```

File: README.md [hunk 3]
```diff
@@ -3,3 +3,204 @@
...
+  model: 'x1024',
+  mass: 1600,
+  enginePower: 135,
+  fuelCapacity: 59,
+});
+```
+
+### array
+
+Generates schema that will match all values specifiec in the schemas in the array.
```
Comment (mmiszy) [inline README.md:158]:
```suggestion
Generates schema that will match all values specified in the schemas in the array.
```

File: README.md [hunk 4]
```diff
@@ -3,3 +3,204 @@
...
+
+If you use nil modifier, output type will be `undefined | null | InferredType`.
+
+```js
+const nilRoleSchema = nil(oneOf(['User', 'Admin']));
+const nilRoleValidator = validate(nilRoleSchema);
+
+// Returns 'User', the output type is undefined | null | 'User' | 'Admin'
+const role = nilRoleValidator('User');
+```
```
Comment (mmiszy) [inline README.md:206]: It would be nice to mention that null/undefined also pass the validation

**Review by mmiszy — CHANGES_REQUESTED** [review]

File: README.md [hunk 5]
```diff
@@ -3,3 +3,223 @@
...
+const stringSchema = string();
+const stringValidator = validate(stringSchema);
+
+// Return 'Micheal'
+const micheal = stringValidator('Micheal');
+```
+
+### number
+
+Generates schema that will match number.
```
Comment (AdamSiekierski) [inline README.md:90]:
```suggestion
Generates schema that will match a number.
```

File: README.md [hunk 6]
```diff
@@ -3,3 +3,223 @@
...
+// Returns 'trout'
+const trout = fishValidator('trout');
+
+// Throws validation error
+const salmon = fishValidator('salmon');
+```
+
+### string
+
+Generates schema that will match string.
```
Comment (AdamSiekierski) [inline README.md:78]:
```suggestion
Generates schema that will match a string.
```

File: README.md [hunk 7]
```diff
@@ -3,3 +3,223 @@
...
+
+// If validation is successful returns data with correct type, throws error otherwise
+const validatedPerson = personValidator(mark);
+```
+
+## API
+
+### oneOf
+
+Generates a schema which will match only specified values.
```
Comment (AdamSiekierski) [inline README.md:63]:
```suggestion
Generates a schema which will match only the specified values.
```

File: README.md [hunk 8]
```diff
@@ -3,3 +3,223 @@
...
+const numberSchema = number();
+const numberValidator = validate(numberSchema);
+
+// Return 3.14
+const pi = numberValidator(3.14);
+```
+
+### boolean
+
+Generates schema that will match boolean.
```
Comment (AdamSiekierski) [inline README.md:102]:
```suggestion
Generates schema that will match a boolean.
```

File: README.md [hunk 9]
```diff
@@ -3,3 +3,223 @@
...
+const booleanSchema = boolean();
+const booleanValidator = validate(booleanSchema);
+
+// Returns true
+const loading = booleanValidator(true);
+```
+
+### date
+
+Generates schema that will match date.
```
Comment (AdamSiekierski) [inline README.md:114]:
```suggestion
Generates schema that will match a JavaScript Date object.
```

File: README.md [hunk 10]
```diff
@@ -3,3 +3,223 @@
...
+
+// Returns Date Tue Jan 12 2021 22:00:20 GMT+0100 (Central European Standard Time)
+const annieBirthday = dateValidator(
+  new Date('Tue Jan 12 2021 22:00:20 GMT+0100 (Central European Standard Time)'),
+);
+```
+
+### object
+
+Generates schema that will match object.
```
Comment (AdamSiekierski) [inline README.md:128]:
```suggestion
Generates schema that will match a object.
```

File: README.md [hunk 11]
```diff
@@ -3,3 +3,223 @@
...
+  model: 'x1024',
+  mass: 1600,
+  enginePower: 135,
+  fuelCapacity: 59,
+});
+```
+
+### array
+
+Generates schema that will match all values specified in the schemas in the array.
```
Comment (AdamSiekierski) [inline README.md:158]:
```suggestion
Generates a schema that will match all the values specified by the schemas passed in the array.
```

**Review by AdamSiekierski — CHANGES_REQUESTED** [review]

File: README.md [hunk 12]
```diff
@@ -3,3 +3,223 @@
...
+  - [object](#object)
+  - [array](#array)
+  - [Modifiers](#modifiers)
+    - [nullable](#nullable)
+    - [optional](#optional)
+    - [nil](#nil)
+
+## Introduction
+
+@typeofweb/schema is lightweight, extensible library for data validation.
```
Comment (mmiszy) [inline README.md:27]:
```suggestion
@typeofweb/schema is a lightweight and extensible library for data validation with full support for TypeScript!
```

File: README.md [hunk 13]
```diff
@@ -3,3 +3,223 @@
...
+
+## Installation
+
+Install package with npm:
```
Comment (mmiszy) [inline README.md:31]:
```suggestion
Install the package with npm:
```

File: README.md [hunk 14]
```diff
@@ -3,3 +3,223 @@
...
+
+`npm install @typeofweb/schema`
+
+Install package with yarn:
```
Comment (mmiszy) [inline README.md:35]:
```suggestion
or install the package with yarn:
```

File: README.md (hunk 7 above)
Comment (mmiszy) [reply README.md:63]: To chyba nie jest poprawnie? :P

Comment (mmiszy) [inline README.md:63]:
```suggestion
Generates a schema which matches only specified values.
```

File: README.md [hunk 15]
```diff
@@ -3,3 +3,223 @@
...
+Generates a schema which will match only specified values.
+
+```js
+const fishSchema = oneOf(['trout', 'catfish']);
+const fishValidator = validate(fishSchema);
+
+// Returns 'trout'
+const trout = fishValidator('trout');
+
+// Throws validation error
```
Comment (mmiszy) [inline README.md:72]:
```suggestion
// Throws ValidationError
```

File: README.md (hunk 6 above)
Comment (mmiszy) [inline README.md:78]:
```suggestion
Creates a schema that matches strings.
```

File: README.md (hunk 5 above)
Comment (mmiszy) [inline README.md:90]:
```suggestion
Creates a schema that matches numbers.
```

File: README.md [hunk 16]
```diff
@@ -3,3 +3,223 @@
...
+
+### number
+
+Generates schema that will match number.
+
+```js
+const numberSchema = number();
+const numberValidator = validate(numberSchema);
+
+// Return 3.14
```
Comment (mmiszy) [inline README.md:96]:
```suggestion
// Returns 3.14
```

File: README.md (hunk 8 above)
Comment (mmiszy) [inline README.md:102]:
```suggestion
Creates a schema that matches booleans.
```

File: README.md (hunk 9 above)
Comment (mmiszy) [inline README.md:114]:
```suggestion
Creates a schema that matches JavaScript `Date` objects.
```

File: README.md (hunk 10 above)
Comment (mmiszy) [inline README.md:128]:
```suggestion
Creates a schema that matches objects of provided shape.
```

File: README.md [hunk 17]
```diff
@@ -3,3 +3,223 @@
...
+const annieBirthday = dateValidator(
+  new Date('Tue Jan 12 2021 22:00:20 GMT+0100 (Central European Standard Time)'),
+);
+```
+
+### object
+
+Generates schema that will match object.
+
+```js
```
Comment (mmiszy) [inline README.md:130]:
```suggestion
```ts
```

File: README.md [hunk 18]
```diff
@@ -3,3 +3,223 @@
...
+
+/* Returns {
+  manufacturer: 'Typeofweb enterprise',
+  model: 'x1024',
+  mass: 1600,
+  enginePower: 135,
+  fuelCapacity: 59,
+} */
+const matthewCar = carValidator({
+  manufacturer: 'Typeofweb enterprise',
```
Comment (mmiszy) [inline README.md:148]:
```suggestion
  manufacturer: 'Type of Web Enterprise',
```

File: README.md (hunk 11 above)
Comment (mmiszy) [inline README.md:158]:
```suggestion
Creates a schema that matches arrays containing values specified by the schemas passed in the array.
```

File: README.md [hunk 19]
```diff
@@ -3,3 +3,223 @@
...
+
+Generates schema that will match all values specified in the schemas in the array.
+
+```js
+const musicGenresSchema = array([string()]);
+const musicGenresValidator = validate(musicGenresSchema);
+
+// Returns ['classical', 'lofi', 'pop']
+const musicGenres = musicGenresValidator(['classical', 'lofi', 'pop']);
+```
```
Comment (mmiszy) [inline README.md:166]: We could use an example with more than one schema inside the array

File: README.md [hunk 20]
```diff
@@ -3,3 +3,223 @@
...
+
+### Modifiers
+
+Modifers are used to customize the schema and change the validation behaviour.
```
Comment (mmiszy) [inline README.md:170]:
```suggestion
Modifiers are used to customize schemas and change their validation behaviour.
```

File: README.md [hunk 21]
```diff
@@ -3,3 +3,223 @@
...
+
+#### nullable
+
+If you use nullable modifier, output type will be `null | InferredType`.
```
Comment (mmiszy) [inline README.md:174]:
```suggestion
If you use the `nullable` modifier, given schema will be extended to also match `null` values, and its output type will be `null | T`.
```

File: README.md [hunk 22]
```diff
@@ -3,3 +3,223 @@
...
+const nullableRoleSchema = nullable(oneOf(['User', 'Admin']));
+const nullableRoleValidator = validate(nullableRoleSchema);
+
+// Returns 'User', the output type is null | 'User' | 'Admin'
+const role = nullableRoleValidator('User');
+
+// It's also fine
+nullableRoleValidator(null);
+
+// Throws validation error
```
Comment (mmiszy) [inline README.md:186]:
```suggestion
// Throws ValidationError
```

File: README.md [hunk 23]
```diff
@@ -3,3 +3,223 @@
...
+nullableRoleValidator();
+nullableRoleValidator(undefined);
+```
+
+### optional
+
+If you use optional modifier, output type will be `undefined | InferredType`.
```
Comment (mmiszy) [inline README.md:193]:
```suggestion
If you use the `optional` modifier, given schema will be extended to also match `undefined` values, and its output type will be `undefined | T`.
```

File: README.md [hunk 24]
```diff
@@ -3,3 +3,223 @@
...
+optionalRoleValidator();
+optionalRoleValidator(undefined);
+
+// Throws validation error
+optionalRoleValidator(null);
+```
+
+### nil
+
+If you use nil modifier, output type will be `undefined | null | InferredType`.
```
Comment (mmiszy) [inline README.md:212]:
```suggestion
If you use the `nil` modifier, given schema will be extended to also match `null` and `undefined` values, and its output type will be `null | undefined | T`.
```

**Review by mmiszy — CHANGES_REQUESTED** [review]

File: README.md [hunk 25]
```diff
@@ -3,3 +3,223 @@
...
+  enginePower: 135,
+  fuelCapacity: 59,
+});
+```
+
+### array
+
+Generates schema that will match all values specified in the schemas in the array.
+
+```js
```
Comment (mmiszy) [inline README.md:160]:
```suggestion
```ts
```

File: README.md [hunk 26]
```diff
@@ -3,3 +3,223 @@
...
+const validatedPerson = personValidator(mark);
+```
+
+## API
+
+### oneOf
+
+Generates a schema which matches only specified values.
+
+```js
```
Comment (mmiszy) [inline README.md:65]:
```suggestion
```ts
```

File: README.md [hunk 27]
```diff
@@ -3,3 +3,223 @@
...
+
+// Throws ValidationError
+const salmon = fishValidator('salmon');
+```
+
+### string
+
+Creates a schema that matches strings.
+
+```js
```
Comment (mmiszy) [inline README.md:80]:
```suggestion
```ts
```

File: README.md [hunk 28]
```diff
@@ -3,3 +3,223 @@
...
+
+// Return 'Micheal'
+const micheal = stringValidator('Micheal');
+```
+
+### number
+
+Creates a schema that matches numbers.
+
+```js
```
Comment (mmiszy) [inline README.md:92]:
```suggestion
```ts
```

File: README.md [hunk 29]
```diff
@@ -3,3 +3,223 @@
...
+
+// Returns 3.14
+const pi = numberValidator(3.14);
+```
+
+### boolean
+
+Creates a schema that matches booleans.
+
+```js
```
Comment (mmiszy) [inline README.md:104]:
```suggestion
```ts
```

File: README.md [hunk 30]
```diff
@@ -3,3 +3,223 @@
...
+
+// Returns true
+const loading = booleanValidator(true);
+```
+
+### date
+
+Creates a schema that matches JavaScript `Date` objects.
+
+```js
```
Comment (mmiszy) [inline README.md:116]:
```suggestion
```ts
```

File: README.md [hunk 31]
```diff
@@ -3,3 +3,223 @@
...
+
+### Modifiers
+
+Modifiers are used to customize schemas and change their validation behaviour.
+
+#### nullable
+
+If you use the `nullable` modifier, given schema will be extended to also match `null` values, and its output type will be `null | T`.
+
+```js
```
Comment (mmiszy) [inline README.md:176]:
```suggestion
```ts
```

File: README.md [hunk 32]
```diff
@@ -3,3 +3,223 @@
...
+// Throws ValidationError
+nullableRoleValidator();
+nullableRoleValidator(undefined);
+```
+
+### optional
+
+If you use the `optional` modifier, given schema will be extended to also match `undefined` values, and its output type will be `undefined | T`.
+
+```js
```
Comment (mmiszy) [inline README.md:195]:
```suggestion
```ts
```

File: README.md [hunk 33]
```diff
@@ -3,3 +3,223 @@
...
+
+// Throws validation error
+optionalRoleValidator(null);
+```
+
+### nil
+
+If you use the `nil` modifier, given schema will be extended to also match `null` and `undefined` values, and its output type will be `null | undefined | T`.
+
+```js
```
Comment (mmiszy) [inline README.md:214]:
```suggestion
```ts
```

**Review by mmiszy — CHANGES_REQUESTED** [review]

**Review by mmiszy — APPROVED** [review]

**Pull request merged**